packages/editor/src/extensions/key-map/move-node.ts

```diff
@@ -0,0 +1,215 @@
+/*
+This file is part of the Notesnook project (https://notesnook.com/)
+
+Copyright (C) 2023 Streetwriters (Private) Limited
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { Editor } from "@tiptap/core";
+import OrderedList from "@tiptap/extension-ordered-list";
+import { Fragment, Node, Slice } from "@tiptap/pm/model";
+import { Selection } from "@tiptap/pm/state";
+import { ReplaceStep } from "@tiptap/pm/transform";
+import { isListActive } from "../../utils/list.js";
+import {
+  findParentNodeClosestToPos,
+  findParentNodeOfType
+} from "../../utils/prosemirror.js";
+import { Blockquote } from "../blockquote/blockquote.js";
+import { BulletList } from "../bullet-list/bullet-list.js";
+import { Callout } from "../callout/callout.js";
+import { CheckList } from "../check-list/check-list.js";
+import { OutlineList } from "../outline-list/outline-list.js";
+import { Table } from "../table/table.js";
+import { TaskListNode } from "../task-list/task-list.js";
+
+type ResolvedNode = {
+  pos: number;
+  start: number;
+  depth: number;
+  node: Node;
+};
+
+function mapChildren<T>(
+  node: Node | Fragment,
+  callback: (child: Node, index: number, frag: Fragment) => T
+): T[] {
+  const array = [];
+  for (let i = 0; i < node.childCount; i++) {
+    array.push(
+      callback(node.child(i), i, node instanceof Fragment ? node : node.content)
+    );
+  }
+  return array;
+}
+
+function resolveNode(editor: Editor) {
+  const { state } = editor;
+  const { $from } = state.selection;
+
+  let targetType = $from.node().type;
+  let currentResolved = findParentNodeOfType(targetType)(state.selection);
+
+  if (isListActive(editor)) {
+    const currentNode = $from.node();
+    const parentNode = $from.node($from.depth - 1);
+    const isFirstParagraph =
+      currentNode.type.name === "paragraph" &&
+      parentNode.firstChild === currentNode;
+
+    // move the entire list item
+    if (isFirstParagraph) {
+      targetType = $from.node($from.depth - 1).type;
+      if (
+        targetType.name === Callout.name ||
+        targetType.name === Blockquote.name
+      ) {
+        targetType = $from.node($from.depth - 2).type;
+      }
+    }
```

Comment:
This is very fragile. It just assumes that list item is the first or second parent but that won't always be the case i.e. what if user has nested callouts or blockquotes?

Reply:
I'm still trying to remember why I have the callout and blockquote specific checks inside `isFirstParagraph` check. We don't need that. I've removed that, and am also making sure that the `parentNode` is a list item before attempting to move the entire list item